typescript/src/bitcoin.ts

```diff
@@ -359,3 +359,62 @@ export function createKeyRing(
 export function computeHash160(text: string): string {
   return hash160.digest(Buffer.from(text, "hex")).toString("hex")
 }
+
+/**
+ * Encodes a public key hash into a P2PKH/P2WPKH address.
+ * @param publicKeyHash - public key hash that will be encoded. Must be an
+ *        unprefixed hex string (without 0x prefix).
+ * @param witness - If true, a witness public key hash will be encoded and
+ *        P2WPKH address will be returned. Returns P2PKH address otherwise
+ * @param network - Network that the address should be encoded for.
+ *        For example, `main` or `testnet`.
+ * @returns P2PKH or P2WPKH address encoded from the given public key hash
+ */
+export function encodeToBitcoinAddress(
+  publicKeyHash: string,
+  witness: boolean,
+  network: string
+): string {
+  return witness
+    ? bcoin.Address.fromWitnessPubkeyhash(
+        Buffer.from(publicKeyHash, "hex")
+      ).toString(network)
+    : bcoin.Address.fromPubkeyhash(Buffer.from(publicKeyHash, "hex")).toString(
+        network
+      )
+}
+
+/**
+ * Decodes P2PKH or P2WPKH address into a public key hash.
+ * @param address - P2PKH or P2WPKH address that will be decoded.
+ * @returns Public key hash decoded from the address. This will be an unprefixed
+ *        hex string (without 0x prefix).
+ */
+export function decodeBitcoinAddress(address: string): string {
+  const addressObject = new bcoin.Address(address)
+  return addressObject.getHash("hex")
+}
+
+/**
+ * Checks if given string is a publick key hash.
+ * @param publicKeyHash - text that will be checked if it's proper pub key hash.
+ * @param witness - If true, it will check whenever it's witness public key
+          hash. It will check it it's "normal" public key hash otherwise.
+ * @returns true if the given public key hash is a proper public key hash, false
+          otherwise.
+ */
+export function isPubkeyhash(publicKeyHash: string, witness: boolean): boolean {
+  try {
+    const addressObject = new bcoin.Address()
+    if (witness) {
+      addressObject.fromWitnessPubkeyhash(Buffer.from(publicKeyHash, "hex"))
+    } else {
+      addressObject.fromPubkeyhash(Buffer.from(publicKeyHash, "hex"))
+    }
+    return witness
+      ? addressObject.isWitnessPubkeyhash()
+      : addressObject.isPubkeyhash()
```

Comment:
When creating this `isPubkeyhash` function I was not sure if we need a `witness` parameter here 🤔 I've decided to put it anyway and just ask what you think.

As far as I understand from the publick key hash we can create either P2PKH address or P2WPKH address, but the public key hash itself does not separate into witness PKH and non-witness PKH 🤔 Looks to me that it's more related to the lib itself and it needs separate functions based on how the `addressObject` was created (`formPubkeyHash` and `fromWitnessPubkeyHash`) which will determinate how we will decode works (different for `P2PKH` address created from `fromPubKeyHash()` and differnt for `P2WPKH` address created from `fromWitnessPubKeyHash()`).

This probably means that we don't need `witness` parameter here at all.

Reply:
Actually, that function doesn't make sense 😄 A public key hash is just a 20-byte hash, nothing more. Moreover, you never know whether an actual ECDSA public key was used to produce that hash. You can produce a 20-byte hash from literally anything. Here you construct a `bcoin` address using the `publicKeyHash` argument. The validation done by `fromWitnessPubkeyhash` and `fromPubkeyhash` just checks if the size is actually 20-byte so, as I said above, any 20-byte hash will pass it and the `bcoin` address will be constructed. The constructed `bcoin` address holds the information about how it was constructed so, if you used `fromWitnessPubkeyhash`/`fromPubkeyhash` then `isWitnessPubkeyhash`/`isPubkeyhash` will always return true.

I understand why you added that function but the only thing we can validate within `validateDepositScriptParameters` is whether the passed `walletPublicKeyHash`/`refundPublicKeyHash` is a 20-byte value. That said, we can remove `isPubkeyhash` and introduce `isPublicKeyHashLength` instead.

Reply:
Fixed in 064f5d9ed9b352a1518b20acf237507c33b7e0b2